File: common/src/main/java/org/sandboxpowered/silica/network/LengthPrepender.java
package org.sandboxpowered.silica.network;

import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.MessageToByteEncoder;

public class LengthPrepender extends MessageToByteEncoder<ByteBuf> {
    public static int getVarIntSize(int i) {
        for(int j = 1; j < 5; ++j) {
            if ((i & -1 << j * 7) == 0) {
                return j;
            }
        }

        return 5;
    }

    @Override
    protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws Exception {
        int i = in.readableBytes();
        int j = getVarIntSize(i);
        if (j > 3) {
            throw new IllegalArgumentException("unable to fit " + i + " into " + 3);
        } else {
            PacketByteBuf packetByteBuf = new PacketByteBuf(out);
            packetByteBuf.ensureWritable(j + i);
            packetByteBuf.writeVarInt(i);
            packetByteBuf.writeBytes(in, in.readerIndex(), i);
        }
    }
}

File: common/src/main/kotlin/org/sandboxpowered/silica/network/LengthPrepender.kt
package org.sandboxpowered.silica.network

import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.MessageToByteEncoder

class LengthPrepender : MessageToByteEncoder<ByteBuf>() {
    override fun encode(ctx: ChannelHandlerContext, buf: ByteBuf, out: ByteBuf) {
        val packetSize = buf.readableBytes()
        val length = getVarIntSize(packetSize)
        require(length <= 3) { "unable to fit $packetSize into 3" }
        val packetByteBuf = PacketByteBuf(out)
        packetByteBuf.ensureWritable(length + packetSize)
        packetByteBuf.writeVarInt(packetSize)
        packetByteBuf.writeBytes(buf, buf.readerIndex(), packetSize)
    }

    private fun getVarIntSize(value: Int): Int {
        for (variable in 1..4) {
            if (value and (-1 shl variable * 7) == 0) {
                return variable
            }
        }
        return 5
    }
}